### PR TITLE
order: ensure runnable tasks are certainly runnable

### PR DESCRIPTION
**Pull request by fjetter**
This fixes a bug in `process_runnables` that caused the `runnable_hull` to be populated with leaf nodes which is not intended to happen. If a leaf node was encountered during `process_runnables` we'd always want it to be executed greedily.

Fixing this bug allows us to eliminate a couple of code branches that are now no longer reachable. It also reveals a bug that was covered up that I think should/will be fixed by https://github.com/dask/dask/pull/11303

While the ordering for `test_run_smaller_sections` is now actually better in terms of max pressure, we are indeed ordering in a not intuitive way and are not considering already loaded root tasks.

### CODE REVIEW

File: dask/order.py [hunk 1]
```diff
@@ -326,6 +326,8 @@ def process_runnables() -> None:
                             if num_needed[current] <= 1:
                                 for k in path:
                                     add_to_result(k)
+                            else:
+                                runnable_hull.discard(current)
```
Comment (fjetter) [inline dask/order.py:330]: the loop invariant this should preserve is that runnable_hull and known_runnable_paths are disjoint

Comment (fjetter) [reply dask/order.py:330]: ok, I stand corrected... that invariant is still not true. I'll investigate...

File: dask/tests/test_order.py [hunk 2]
```diff
@@ -339,6 +339,9 @@ def test_favor_longest_critical_path(abcde):
     assert o[e] > o[b]
 
 
+@pytest.mark.xfail(
+    reason="Second target pick does not include already computed nodes properly"
```
Comment (phofl) [inline dask/tests/test_order.py:343]: This test will become green again after #11303

File: dask/order.py [hunk 3]
```diff
@@ -315,6 +315,8 @@ def process_runnables() -> None:
                 while branches:
                     path = branches.popleft()
                     while True:
+                        # Loop invariant. Too expensive to compute at runtime
+                        # assert not set(known_runnable_paths).intersection(runnable_hull)
```
Comment (fjetter) [inline dask/order.py:319]: this makes me want to introduce a `_validate` kwarg that allows us to enable these kinds of assertions without having them be active at runtime (alternatively via ctx variable)

Comment (phofl) [reply dask/order.py:319]: Yeah this makes sense I think, especially if we keep tinkering

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx

**Pull request merged**